### ISSUE TITLE
Bar overflowing with custom domain and margin

### ISSUE THREAD

**Issue by Lueton** [recharts/recharts#4267]
### Reproduction link
[![Edit on CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/sandbox/recharts-issue-template-forked-v34gq4?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cltfitlao00063j6i3u1anjbg%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cltfitlao00023j6i796ptmkt%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cltfitlao00033j6iq9m5bvg4%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cltfitlao00053j6iq6xfj0lf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cltfitlao00023j6i796ptmkt%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltfitlan00013j6ii1eftm3x%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A35%252C%2522startColumn%2522%253A46%252C%2522endLineNumber%2522%253A35%252C%2522endColumn%2522%253A46%257D%255D%257D%255D%252C%2522id%2522%253A%2522cltfitlao00023j6i796ptmkt%2522%252C%2522activeTabId%2522%253A%2522cltfitlan00013j6ii1eftm3x%2522%257D%252C%2522cltfitlao00053j6iq6xfj0lf%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cltfitlao00043j6idf9nuaoh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522cltfitlao00053j6iq6xfj0lf%2522%252C%2522activeTabId%2522%253A%2522cltfitlao00043j6idf9nuaoh%2522%257D%252C%2522cltfitlao00033j6iq9m5bvg4%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522cltfitlao00033j6iq9m5bvg4%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Steps to reproduce
Open the sandbox

### What is expected?
I would expect the purple bar to stop at the end of the chart. It should not overflow the chart container and respect the margin  

### What is actually happening?
The margin does not get respected and the bar is overflowing 

  
![image](https://github.com/recharts/recharts/assets/42818194/afd63916-2aba-4259-9c7c-a241c889a2fc)

| Environment | Info |
|---|---|
| Recharts | v2.12.2 |
| React | 16.8.6 |
| System | Windows |
| Browser | Edge, Chrome |

**Comment by ckifer** [comment]
interesting, it looks like `top` margin is the issue here. Wonder why this had never been found previously...

Commenting out `top: 100,` "fixes" this, but looks like this is just a long time bug.

@HHongSeungWoo @PavelVanecek fyi if anyone wants to look into this one (during refactoring or otherwise)

**Comment by Lueton** [comment]
> interesting, it looks like `top` margin is the issue here. Wonder why this had never been found previously...
> 
> Commenting out `top: 100,` "fixes" this, but looks like this is just a long time bug.
> 
> @HHongSeungWoo @PavelVanecek fyi if anyone wants to look into this one (during refactoring or otherwise)

Im using Recharts for several years, but I actually found this problem when writing my own chart component, which had the same problem. Was curious how Recharts solved this problem. For me, the problem was about the same. If you have a custom domain (like in my example) or limit, but don't limit the bar size to the given limit it will just overflow. This is not visible until you have a margin because obviously you cant draw outside the svg. But adding a margin makes the overflow visible.

**Comment by ckifer** [comment]
Makes sense. We would probably put a clip on the top of the chart boundary, inclusive of margin. Looks like right now there is no clip (or anything else) and therefore, as you say it just overflows.

**Comment by tannerdolby** [comment]
Setting `allowDataOverflow={true}` on the YAxis makes needClip true in the render method of `Bar.tsx` and the `<clipPath>` is used which hides the overflowing bars.

- https://github.com/recharts/recharts/blob/3.x/src/cartesian/Bar.tsx#L496

<details><summary>Example code with overflow fixed</summary>
<p>


```js
import React, { PureComponent } from "react";
import {
  BarChart,
  Bar,
  Rectangle,
  XAxis,
  YAxis,
  CartesianGrid,
  Tooltip,
  Legend,
} from "recharts";

const data = [
  {
    name: "Page A",
    uv: 4000,
    pv: 2400,
    amt: 2400,
  },
  {
    name: "Page B",
    uv: 3000,
    pv: 1398,
    amt: 2210,
  },
  {
    name: "Page C",
    uv: 2000,
    pv: 9800,
    amt: 2290,
  },
  {
    name: "Page D",
    uv: 2780,
    pv: 3908,
    amt: 2000,
  },
  {
    name: "Page E",
    uv: 1890,
    pv: 4800,
    amt: 2181,
  },
  {
    name: "Page F",
    uv: 2390,
    pv: 3800,
    amt: 2500,
  },
  {
    name: "Page G",
    uv: 3490,
    pv: 4300,
    amt: 2100,
  },
];

export default function BarGraph() {
    return (
    <BarChart
        width={500}
        height={300}
        data={data}
        margin={{
          top: 100,
          right: 30,
          left: 20,
          bottom: 5,
        }}
      >
        <CartesianGrid strokeDasharray="3 3" />
        <XAxis dataKey="name" />
        <YAxis domain={[0, 5000]} type={"number"} dataKey="uv" allowDataOverflow />
        <Tooltip />
        <Legend />
        <Bar
          dataKey="pv"
          fill="#8884d8"
          activeBar={<Rectangle fill="pink" stroke="blue" />}
        />
        <Bar
          dataKey="uv"
          fill="#82ca9d"
          activeBar={<Rectangle fill="gold" stroke="purple" />}
        />
      </BarChart>
    );
}
```
</p>
</details>
